### ISSUE TITLE
[bug] Applying updates can result in unnecessary additional updates on multi-value records

### ISSUE THREAD

**Issue by sshipway** [ngoduykhanh/PowerDNS-Admin#735]
This is a difficult one and I'm still trying to reliably replicate it. 

If you have multi-values RRSets, such as multiple NS records in a delegation, then Powerdns-admin can sometimes do an unnecessary update of these when you update other records.

For example, if you have:
```
foo 3600 IN NS  ns0
foo 3600 IN NS ns1
```
Then you add
`foo 3600 IN TXT "foo"`
... this should result in an **add_rrsets** for the 'foo'.
However, _sometimes_, it also results in a **del_rrsets** for the NS records, and an **add_rrsets** for the same records but in a _different order_.

Example from logs:
```
{
   "domain" : "testdomain.com",
   "add_rrests" : [
      {
         "comments" : [],
         "name" : "365.nz.testdomain.com.",
         "records" : [
            {
               "content" : "20 shared2.nz.testdomain.com.",
               "disabled" : false
            },
            {
               "content" : "10 shared1.nz.testdomain.com.",
               "disabled" : false
            }
         ],
         "ttl" : 7200,
         "type" : "MX",
         "changetype" : "DELETE"
      }
   ],
   "del_rrests" : [
      {
         "name" : "365.nz.testdomain.com.",
         "type" : "MX",
         "ttl" : 7200,
         "records" : [
            {
               "content" : "10 shared1.nz.testdomain.com.",
               "disabled" : false
            },
            {
               "content" : "20 shared2.nz.testdomain.com.",
               "disabled" : false
            }
         ],
         "comments" : [],
         "changetype" : "REPLACE"
      }
   ]
}
```
As you can see, no net change, just the ordering of the components changed.  When this happens, it affects _every_ multivalued record set and so ther update is huge.

The net effect of this is, of course, nothing - until you try and save an illegal record, then something goes wrong resulting in the **del_rrsets** being processed but not the **add_rrsets**.  This is, as you can imaging, not something you want.  A fault today erased every multi-valued record in our zone, and logged nothing.

**Comment by sshipway** [comment]
I have now proved this.  Created a new zone, added 3 text records for name 'a', all with different text. Applied the changes, refreshed the page.  Sorted the columns in different orders.  Created a new TXT record 'b'.  The logs show the update write 'b', but also deleted and re-created 'a' with the records in a different order.

**Comment by sshipway** [comment]
I suppose a fix would be to sort the `records` array in the recordsets before comparing, but I cant find in the code where this comparison happens!  Only a compare() function that seems to check for nothing other than if the record already exists.  I need someone who understands php.

**Comment by sshipway** [comment]
Would it work to add
```
            # Merge rrsets into one
            $sortby = array_column($rrsets, 'records');
            array_multisort($sortby, SORT_DESC, $rrsets);
```
in `merge_rrsets()` of records.py -- to make the returned rrsets always merged in a determined order?

It would probably then be necessary to do the same sorting in `get_rrsets()` somehow, maybe by walking through the `jdata['rrsets']` and sorting the rsets where appropriate 
```
            rrsets=jdata['rrsets']
            for r in rrsets[1:]:
                if count(r['records'])<2:
                    sortby = array_column(r, 'records');
                    array_multisort(sortby, SORT_DESC, r['records']);
```

Again, I'm not a php person, so would like to have some feedback on this.

**Comment by sshipway** [comment]
OK, I have a fix for this now; however it requires also that the problem with comments is fixes as well (IE that the comments array is either empty or has the same number of elements as the records array).  I will make a PR shortly

**Comment by sshipway** [comment]
The fix for this depends on #736 #720 and PR #684 , all of which address the problem with index errors when a Recordset has comments on some but not all of the entries.

The main problem is that comments are not handled properly when some but not all of a recordset members have a comment...

**Comment by sshipway** [comment]
@ngoduykhanh I've now rolled out our commit that fixes this ( in PR #739 ) to our systems here, as the risk of damage to our data from the bug was too high.  It does seem to fix both this issue and the other 3 I linked above, at least as far as my few hours of testing could determine.  Please consider prioritising merging so that we can return to an official release.

**Comment by ngoduykhanh** [comment]
I've merged #739. Thanks @sshipway .
